### ISSUE TITLE
How to quantify ssd_mobilenet_v1_coco model and toco to .tflite ?

### ISSUE THREAD

**Issue by WenguoLi** [tensorflow/tensorflow#18829]
## System information
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- TensorFlow installed from (source or binary): Source
- TensorFlow version (use command below): 1.8.0
- Python version:2.7.12
- Bazel version (if compiling from source): 0.12.0
- CUDA/cuDNN version: cuda-9.0/7.0
- GPU model and memory: GeForce GTX 1080/8105MiB
- Exact command to reproduce: N/A
- Phone: Moto G4 Play (need to use Camera, not Camera2)

Hi,  guys,
I am trying to quantify ssd_mobilenet_v1 using tensorflow object detection api, according to @coutner [post ](https://github.com/tensorflow/tensorflow/issues/18342).
First, I replace all the graph_hook_fn in
 [trainer.py](https://github.com/tensorflow/models/blob/b6bcc450b981eba721ee2760c92d87da86900988/research/object_detection/trainer.py) tf.contrib.quantize.create_training_graph and enable fused batch norm in [ssd_mobilenet_v1_feature_extractor.py](https://github.com/tensorflow/models/blob/b6bcc450b981eba721ee2760c92d87da86900988/research/object_detection/models/ssd_mobilenet_v1_feature_extractor.py#L114)
After training, I used the following script to transform the model.

1. when running optimize_for_inference, some warning had arisen.
```
$ bazel run -c opt tensorflow/python/tools/optimize_for_inference -- \
>     --input=$DETECT_PB --output=$STRIPPED_PB --frozen_graph=True \
>     --input_names=Preprocessor/sub --output_names=concat,concat_1 \
>     --toco_compatible=True \
>     --alsologtostderr
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_2_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_3_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_4_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_5_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_6_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_7_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_8_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_9_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_10_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_11_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_12_depthwise/BatchNorm/FusedBatchNorm'
WARNING:tensorflow:Didn't find expected Conv2D input to 'FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_13_depthwise/BatchNorm/FusedBatchNorm'
```
When Running toco,
```
$ bazel run tensorflow/contrib/lite/toco:toco -- 
>   --input_file=$STRIPPED_PB \
>   --output_file=$DETECT_FB \
>   --input_format=TENSORFLOW_GRAPHDEF \
>   --output_format=TFLITE \
>   --inference_type=QUANTIZED_UINT8 \
>   --input_shapes=1,300,300,3 \
>   --input_arrays=Preprocessor/sub \
>   --output_arrays=concat,concat_1 \
>   --std_values=128 \
>   --mean_values=127 \
>   --dump_graphviz=/tmp
2018-04-24 18:51:51.645315: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 253 operators, 450 arrays (0 quantized)
2018-04-24 18:51:51.686011: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 253 operators, 450 arrays (0 quantized)
2018-04-24 18:51:51.774563: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 62 operators, 169 arrays (1 quantized)
2018-04-24 18:51:51.775525: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 62 operators, 169 arrays (1 quantized)
2018-04-24 18:51:51.776094: F tensorflow/contrib/lite/toco/tooling_util.cc:1464] Array FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/Relu6, which is an input to the DepthwiseConv operator producing the output array FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_depthwise/Relu6, is lacking min/max data, 
which is necessary for quantization. Either target a non-quantized output format, or change the input 
graph to contain min/max information, or pass --default_ranges_min= and --default_ranges_max= if you 
do not care about the accuracy of results.
Aborted (core dumped)
```
Is there anyone who can help ? @andrewharp ,
thanks.

**Comment by andrehentz** [comment]
Can you try --inference_type=FLOAT?

**Comment by GPhilo** [comment]
@andrehentz doesn't that disable quantization? I thought the whole point of @WenguoLi 's question was to _quantize_ in tflite...

**Comment by zhyj3038** [comment]
@MohammadMoradi   hello, I face the same pronlem, "Cannot convert between a TensorFlowLite buffer with 270000 bytes and a ByteBuffer with 1080000 bytes" .  have you solved it ?

**Comment by MohammadMoradi** [comment]
@zhyj3038 hello, the problem is that when using float model, you must allocate 4 * 270000 = 1080000 because the float is 32bit and the quantized version of model is 8bit. You must change the input allocated buffer to match your model type.

**Comment by zhyj3038** [comment]
@MohammadMoradi    thanks very much .   I am not good at java and android , you help me a lot !

**Comment by alokranjan007** [comment]
@MohammadMoradi @WenguoLi @zhyj3038 
Hello all,

I am facing the same problem when I am using `inference_type=FLOAT`. Do you guys solved the problem? 
Where should I change the value for bytebuffer? I am not getting how should one allocate 4*270000 when using float buffer. 

Please share your suggestion.

Thank you

**Comment by jass-singh** [comment]
@MohammadMoradi @WenguoLi  @zhyj3038 
Hello , 
any update on @alokranjan007 problem ?
I am also having the same issue.
Thanks!

**Comment by zhyj3038** [comment]
@jass-singh    you can not use float, but you must quantize the model.

**Comment by jass-singh** [comment]
@zhyj3038 i tried that too using 
--inference_type=QUANTIZED_UINT8
but then i am not able to create .tflite  : 


2018-08-24 13:16:55.166755: F tensorflow/contrib/lite/toco/tooling_util.cc:1619] Array FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_0/Relu6, which is an input to the DepthwiseConv operator producing the output array FeatureExtractor/MobilenetV1/MobilenetV1/Conv2d_1_depthwise/Relu6, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.
Aborted (core dumped)

**Comment by zhyj3038** [comment]
I use mobilenet v2

**Comment by alokranjan007** [comment]
@jass-singh @zhyj3038 Hello,

A fix: I would recommend that you use`inference_type=FLOAT` for the conversion. However, do change DetectorActivity.java file in the tensorflow/contrib/lite/examples/android/app/src/main/java as below:
private static final boolean TF_OD_API_IS_QUANTIZED = false; (It should be true by default).

After this save your file and try to build the apk. Let me know if it works :)


Cheers!!

**Comment by jass-singh** [comment]
@alokranjan007 
I did but got error : 
Caused by: java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite buffer with 1080000 bytes and a ByteBuffer with 270000 bytes.

Where should i make changes to handle bytes problem?

**Comment by alokranjan007** [comment]
@jass-singh Could you share the detailed steps of converting frozen graph into .tflite format. Just want to check if your are following in right direction. Which model you are using?

Thank you

**Comment by jass-singh** [comment]
> I am using Mobilenet model ssd_mobilenet_v1_coco_2017_11_17 , retraining on faces.

export PYTHONPATH=$PYTHONPATH:`pwd`:`pwd`/slim
export CONFIG_FILE=/home/jaskaran/Git_Download_Libraries/face_detection_inference_graph_2018_08_23/pipeline.config
export CHECKPOINT_PATH=/home/jaskaran/Git_Download_Libraries/face_detection_inference_graph_2018_08_23/model.ckpt
export OUTPUT_DIR=/home/jaskaran/Git_Download_Libraries/face_detection_inference_graph_2018_08_23/

> Further, creating .pb using pipeline.config,

python object_detection/export_tflite_ssd_graph.py \
--pipeline_config_path=$CONFIG_FILE \
--trained_checkpoint_prefix=$CHECKPOINT_PATH \
--output_directory=$OUTPUT_DIR \
--add_postprocessing_op=true


> Further, made changes in DetectorActivity.java file in the tensorflow/contrib/lite/examples/android/app/src/main/java as below:
private static final boolean TF_OD_API_IS_QUANTIZED = false;


> At last ,

bazel run -c opt tensorflow/contrib/lite/toco:toco -- \
--input_file=$OUTPUT_DIR/tflite_graph.pb \
--output_file=$OUTPUT_DIR/custom_mob_2018_08_23_v2.tflite \
--input_shapes=1,300,300,3 \
--input_arrays=normalized_input_image_tensor \
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3'  \
--inference_type=FLOAT \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops

**Comment by alokranjan007** [comment]
@jass-singh Assuming that you have already trained you model and have set the python path. I would suggest to follow below steps for inferncin.

Step 1.
`python export_tflite_ssd_graph.py \
--pipeline_config_path=/home/username/tensorflow-master/models/research/object_detection/path to .config file \
--trained_checkpoint_prefix=/home/username/tensorflow-master/models/research/object_detection/path to the trained model ckpt point(your total no of steps after the training) \
--output_directory=/home/username/tensorflow-master/models/research/object_detection/any folder name \
--add_postprocessing_op=true`
This will generate tflite_graph.pb file and tflite_graph.pbtxt

Step 2:
`bazel run -c opt tensorflow/contrib/lite/toco:toco -- \
--input_file=/home/username/tensorflow-master/models/research/object_detection/customized_model/path to tflite_graph.pb \
--output_file=/home/username/tensorflow-master/models/research/object_detection/your any folder name/detect.tflite \
--input_shapes=1,300,300,3 \
--input_arrays=normalized_input_image_tensor \
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3'  \
--inference_type=FLOAT \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops`

After this detect.tflite file will be generated then you can use this to port on device.

**Comment by jass-singh** [comment]
Hello guys,
@alokranjan007 @MohammadMoradi @WenguoLi @zhyj3038
i found one solution ,

After creating .pb file ,
add 
--default_ranges_min=0 \
--default_ranges_max=255 \
along with 
--inference_type=QUANTIZED_UINT8 \
It will create .tflite successfully and will work on mobile without causing any bytes issue.


Eg:-

bazel run -c opt tensorflow/contrib/lite/toco:toco -- \
--input_file=$OUTPUT_DIR/tflite_graph.pb \
--output_file=$OUTPUT_DIR/custom_mobilenetv3.tflite \
--input_shapes=1,300,300,3 \
--input_arrays=normalized_input_image_tensor \
--output_arrays='TFLite_Detection_PostProcess','TFLite_Detection_PostProcess:1','TFLite_Detection_PostProcess:2','TFLite_Detection_PostProcess:3' \
--inference_type=QUANTIZED_UINT8 \
--default_ranges_min=0 \
--default_ranges_max=255 \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops

Thanks,

**Comment by alokranjan007** [comment]
@jass-singh That's great that it is working now. I have not tried the above mentioned step. I will check.

Do you have any details how this "default_range_min and max" is solving the problem? Is there any link for detailed insights?

Cheers,

**Comment by chenyuZha** [comment]
@jass-singh  Have you add other options when you run `bazel` to convert .pb to .tflite? 
Because I have  also retrained my graph from the same checkpoint **ssd_mobilenet_v1_coco_2017_11_17** as yours.
So the model is supposed to be a **float model**. 
If I convert to .pb to .tflite without create the QUANTIZED_MODEL, I can get my .tflite successfully , but then I'll have the issue for Android studio (even I set **TF_OD_API_IS_QUANTIZED=false**), the app is crashed each time..
But When I tried your method to convert , I had the following error:
`2018-10-03 18:49:12.586196: F tensorflow/contrib/lite/toco/graph_transformations/quantize.cc:560] Check failed: is_rnn_state_array 
[1]    28561 abort (core dumped)  bazel run -c opt //tensorflow/contrib/lite/toco:toco --   `
Also could you tell me which branch you used when you did the covertion? ( I tried master and r1.9, the master could not even convert .pb to .tflite without option QUANTIZED_MODEL..)

**Comment by switchedfabric** [comment]
@WenguoLi I see you closed on the recommendation to try "try --inference_type=FLOAT".  

I have the exact same issue: it works for a float model, but not for a quantized model (--inference_type=QUANTIZED_UINT8).

The float model is too slow for me, and adding --default_ranges_min=0 and --default_ranges_max=255 makes the model unusable because of the low accuracy.

I'm retraining a custom image set from a quantized model [ssd_mobilenet_v1_quantized_coco](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) from the model zoo.

@andrehentz any ideas how to go forward?

**Comment by anonym24** [comment]
@switchedfabric  did you solve it? How convert pb to quant tflite?

**Comment by anonym24** [comment]
>  Array MobilenetV1/MobilenetV1/Conv2d_0/Relu6, which is an input to the DepthwiseConv operator producing the output array MobilenetV1/MobilenetV1/Conv2d_1_depthwise/Relu6, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.

It says that it is better to do "_run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information_"

But all I have is .pb file generated by retrain.py (for classification weights), there is no any checkpoints

**Comment by ignvinay** [comment]
Array MobilenetV1/MobilenetV1/Conv2d_0/Relu6, which is an input to the DepthwiseConv operator producing the output array MobilenetV1/MobilenetV1/Conv2d_1_depthwise/Relu6, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.

I am also getting same problem.
When i run with default_ranges_min=0 and default_ranges_max=255, the results are way off.
Any idea how we can solve this ??

**Comment by Iorek73** [comment]
@ignvinay , from the [here](https://www.tensorflow.org/lite/convert/cmdline_examples), it's recommended that the default_ranges_max=6 due to the Relu6 function.

**Comment by cpark40** [comment]
> Can you try --inference_type=FLOAT?

thank you so much! you saved me

**Comment by duvictor** [comment]
ihave faced a lot problems to get my model working using tensorflow lite. My thanks to this perfect forum. After read some books and undestand the all the kind process with tensorflow, i finally get my model working in a android envoriment

**Comment by mahamatnoumai** [comment]
@duvictor Hi there, am glad that you got your model working, do you mind sharing your process, and difficulties, what i can see now, we are all facing these issues right now. the issues of converting tf1 model to **QUANTIZED_UINT8**, and with **inference_type=FLOAT** it convert successfully, but when you run on the model, it's give error even though **TF_OD_API_IS_QUANTIZED = false;** Thanks.

**Comment by mahamatnoumai** [comment]
> @WenguoLi I see you closed on the recommendation to try "try --inference_type=FLOAT".
> 
> I have the exact same issue: it works for a float model, but not for a quantized model (--inference_type=QUANTIZED_UINT8).
> 
> The float model is too slow for me, and adding --default_ranges_min=0 and --default_ranges_max=255 makes the model unusable because of the low accuracy.
> 
> I'm retraining a custom image set from a quantized model [ssd_mobilenet_v1_quantized_coco](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md) from the model zoo.
> 
> @andrehentz any ideas how to go forward?

Hi, how did it go after you retrained the custom quantized model [ssd_mobilenet_v1_quantized_coco]
